### DISCUSSION TITLE
Remove attachments in body

### DISCUSSION THREAD

**Discussion by GiaHungTran** [rspamd/rspamd#5229] [Q&A] [answered]
Hi all,

I am using mailcow as a mail server and they have rspamd as a spam filter. I am still new with both. The requirement is that I need to remove all attachments in the incoming email before sending it to the recipient. As I understand, I can write a Lua script to remove the attachments but I couldn't load the script into rspamd. Somehow it is missing a lot of necessary documents that show me how to write my own rule correctly. Can someone please help me?

thank you very much!

`
rspamd_config:register_symbol({
    name = "REMOVE_ATTACHMENTS",
    type = "postfilter",
    priority = 10,
    callback = function(task)
        local parts = task:get_parts()
        if not parts then
            return
        end
        for _, part in ipairs(parts) do
            if part and part.get_header then
                local disposition = part:get_header("Content-Disposition") or ""
                if disposition:match("attachment") or disposition:match("inline") then
                    part:remove() -- Safely remove the part
                end
            end
        end
    end,
})
`

**Comment by GiaHungTran** [comment]
I took a step forward.
in https://rspamd.com/doc/lua/rspamd_mimepart.html there is no function to remove the attachments. So I use
https://rspamd.com/doc/lua/rspamd_task.html

to completely rewrite the email with this code

`local rspamd_logger = require "rspamd_logger"
local rspamd_text = require "rspamd_text"

rspamd_config.REMOVE_ATTACHMENTS = function(task)
    -- Log that the function has started
    rspamd_logger.infox("Starting REMOVE_ATTACHMENTS function")

    -- Get the raw message content
    local raw_message = task:get_message()
    if not raw_message then
        rspamd_logger.errx("No message content found in the task")
        return
    end

    -- Convert rspamd_text to Lua string
    local message = raw_message:str()
    rspamd_logger.infox("Successfully retrieved the raw message")

    -- Extract the boundary from the Content-Type header
    local boundary_pattern = "boundary=([\"']?)([^\"';]+)%1"
    local boundary = message:match(boundary_pattern)

    if not boundary then
        rspamd_logger.errx("No MIME boundary found; skipping attachment removal")
        return
    end

    rspamd_logger.infox("Found MIME boundary: %1", boundary)

    -- Process the MIME parts
    local new_message = {}
    local boundary_marker = "--" .. boundary
    local in_attachment = false

    for line in message:gmatch("[^\r\n]+") do
        if line:find(boundary_marker) then
            in_attachment = false -- Reset on new boundary
        end

        if line:find("Content-Disposition: attachment") or line:find("filename=") then
            rspamd_logger.infox("Attachment detected; skipping content")
            in_attachment = true
        end

        if not in_attachment then
            table.insert(new_message, line)
        end
    end

    -- Log the end of the processing loop
    rspamd_logger.infox("Finished processing message parts")

    -- Log the last 500 characters of the concatenated message
    local last_500_message = string.sub(concatenated_message, -500)
    rspamd_logger.infox("Last 500 characters of the processed message:\n%1", last_500_message)

    -- Update the task with the modified message
    local result = task:set_message(rspamd_text.fromstring(concatenated_message))

    -- Reparse the updated message
    task:process_message()

    -- Retrieve and log the updated message from the task
    local updated_message = task:get_message():str()
    local last_500_updated = string.sub(updated_message, -500)
    rspamd_logger.infox("Message after set_message and re-parsing:\n%1", last_500_updated)

    rspamd_logger.infox("Attachments removed and message updated successfully")
end`


in the end I can see that the logs of the updated emails no longer contain attachments, but in the recipient's inbox the emails still show with attachments. Does anyone know why?

**Comment by GiaHungTran** [comment]
For those still looking for a solution. I wrote my own lua script to remove the attachments. This will rewrite the entire message without the attachment part.

`local rspamd_logger = require "rspamd_logger"
local rspamd_text = require "rspamd_text"

rspamd_config:register_post_filter(function(task)
    rspamd_logger.infox("Starting EML attachment removal and reconstruction")

    -- Get the "From" header
    local from_header = task:get_header("From") or "unknown"
    local clean_from = from_header:gsub("[^%w@.-]", "_")

    -- Get the full MIME message
    local raw_message = task:get_message()
    if not raw_message then
        rspamd_logger.errx("No raw message found; aborting process")
        return
    end

    -- Convert raw message to string
    local message = raw_message:str()
    rspamd_logger.infox("Retrieved raw message for processing")

    -- Extract and log all boundaries
    local boundaries = {}
    for boundary in message:gmatch('boundary="?([^"\r\n;]+)"?') do
        rspamd_logger.infox("Found boundary: %1", boundary)
        table.insert(boundaries, boundary)
    end

    -- Use the first boundary as the top-level boundary
    local top_boundary = boundaries[1]
    if not top_boundary then
        rspamd_logger.errx("No MIME boundary found; skipping processing")
        return
    end
    rspamd_logger.infox("Using top-level boundary: %1", top_boundary)

    -- Process MIME parts
    local new_message = {}
    local boundary_marker = "--" .. top_boundary
    local closing_marker = boundary_marker .. "--"
    local in_attachment = false

    for line in message:gmatch("(.-)\r?\n") do
        if line == boundary_marker then
            if in_attachment then
                rspamd_logger.infox("Finished skipping attachment content")
                table.insert(new_message, "")
            end
            in_attachment = false -- Reset
            table.insert(new_message, line) -- Keep the boundary
        elseif line == closing_marker then
            rspamd_logger.infox("Closing boundary detected: %1", line)
            table.insert(new_message, "")
            table.insert(new_message, line)
        elseif line:lower():find("content%-disposition: attachment") or line:find("filename=") then
            rspamd_logger.infox("Attachment header detected: %1", line)
            in_attachment = true -- Start skipping attachment content
            -- remove remaining headers
            while (string.match(new_message[#new_message], "^[%w%-]+:%s?.*"))
            do
                table.remove(new_message, nil)
            end
        elseif not in_attachment then
            table.insert(new_message, line)
        end
    end

    -- Concatenate the new message
    local concatenated_message = table.concat(new_message, "\r\n")

    -- Save the modified message to a file
    local final_file = "/tmp/cleaned_" .. clean_from .. ".eml"
    local file = io.open(final_file, "w")
    if file then
        file:write(concatenated_message)
        file:close()
        rspamd_logger.infox("Final cleaned email saved to %1", final_file)
    else
        rspamd_logger.errx("Failed to write final email to %1", final_file)
    end

    -- Replace the original message with the reconstructed content
    local success = task:set_message(rspamd_text.fromstring(concatenated_message))
    if success then
        task:process_message()
        rspamd_logger.infox("Successfully removed attachments and replaced email content")
    else
        rspamd_logger.errx("Failed to update the message content")
    end
end)`

**Reply by vstakhov** [reply]
You have probably made all possible mistakes in your script, I don't event dare to review individual ones. I'd suggest to check `rspamadm mime` tool and probably ask some LLM to write what you want based on that.

**Reply by GiaHungTran** [reply]
> You have probably made all possible mistakes in your script, I don't event dare to review individual ones. I'd suggest to check `rspamadm mime` tool and probably ask some LLM to write what you want based on that.

hi, yes I know it's not a good idea to go through each line one by one and just remove the attachments. I wish someone could tell me what is the best function in rspamd to use for this as I couldn't find any useful function for it. What is the critical point you can see here in my approach? I mean if nothing too bad happens I can still use this. BTW I used rspamadmin to revalidate the final message and it worked as desired. What is the abbreviation for LLM?

**Reply by vstakhov** [reply, marked as answer]
https://github.com/rspamd/rspamd/pull/5233
